### PR TITLE
DRDT-107 : Porting Nativo Plugin

### PR DESCRIPTION
**Pull request by akhan3051**

### CODE REVIEW

**Comment by fariasf** [comment]
Created new staging environment for pr-76. Here are some useful links:

[WordPress Admin](https://pr-76-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-76-drdt.pantheonsite.io/)
[Article](https://pr-76-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-76-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-76-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-76-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2553/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Updated staging environment for pr-76. Here are some useful links:

[WordPress Admin](https://pr-76-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-76-drdt.pantheonsite.io/)
[Article](https://pr-76-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-76-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-76-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-76-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2563/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-76 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [97](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/pr-76/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/pr-76/article/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [98](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/pr-76/listicle/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/pr-76/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/pr-76/hub/lighthouse.report.html) | [95](https://circleci.com/gh/fariasf/drdt-test/2562/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Review by fariasf — CHANGES_REQUESTED** [review]
This PR seems to have issues with the `.gitignore`. It includes a bunch of plugins not related to Nativo. Please fix before review. Thanks!

File: web/wp-content/plugins/tmbi-nativo/inc/class-tmbi-nativo-in-articles.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	 * @param string $word_count for number of words.
+	 * @param string $paragraph_number for number of paragraphs.
+	 * @param string $total_paragraphs for total paragraphs.
+	 * @return array()
+	 */
+	public function inject_nativo( $word_count, $paragraph_number, $total_paragraphs ) {
+		$flag              = false;
+		$nativo_ad_tag     = '';
+		$is_last_paragraph = $paragraph_number === $total_paragraphs;
+		if ( 3 === $paragraph_number && 100 >= $word_count ) {
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-nativo/inc/class-tmbi-nativo-in-articles.php]: The `3` here seems to reference the `CADENCE` constant?

File: web/wp-content/plugins/tmbi-nativo/inc/class-tmbi-nativo-in-articles.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			$nativo_ad_tag = $this->get_in_article_ad_markup( '1' );
+			$flag          = true;
+		} elseif ( 3 > $paragraph_number && 100 >= $word_count && ! $is_last_paragraph ) {
+			$nativo_ad_tag = $this->get_in_article_ad_markup( '1' );
+			$flag          = true;
+		}
+		return array(
+			'flag'       => $flag,
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-nativo/inc/class-tmbi-nativo-in-articles.php]: There's no need to return a `flag` - returning either the `$nativo_ad_tag` or `null`/`false` should be enough.

File: web/wp-content/plugins/tmbi-nativo/class-tmbi-nativo.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+class TMBI_Nativo {
+	const VERSION            = '1.0';
+	const NATIVO_CUSTOM_TYPE = 'nativo_custom_post';
+
+	/**
+	 * Add hooks and filters for this module.
+	 *
+	 * @action init
+	 */
+	public function __construct() {
```
Comment (fariasf) [inline web/wp-content/plugins/tmbi-nativo/class-tmbi-nativo.php:34]: I think there's no need to instantiate the class if all the action happens here in the constructor. I've seen how instantiating the class results in extra opcode, compared to just adding the filters outside the class (and referencing the functions like `array( 'TMBI_Nativo', 'register_nativo_post_type' )`) but I'm not sure if it has a performance impact. I'll create a ticket to review and confirm!

**Comment by fariasf** [comment]
Updated staging environment for pr-76. Here are some useful links:

[WordPress Admin](https://pr-76-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-76-drdt.pantheonsite.io/)
[Article](https://pr-76-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-76-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-76-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-76-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2692/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-76 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [95](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/pr-76/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/pr-76/article/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [97](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/pr-76/listicle/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/pr-76/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/pr-76/hub/lighthouse.report.html) | [94](https://circleci.com/gh/fariasf/drdt-test/2691/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Comment by fariasf** [comment]
Updated staging environment for pr-76. Here are some useful links:

[WordPress Admin](https://pr-76-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-76-drdt.pantheonsite.io/)
[Article](https://pr-76-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-76-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-76-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-76-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2701/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-76 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [95](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/pr-76/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/pr-76/article/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [98](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/pr-76/listicle/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/pr-76/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/pr-76/hub/lighthouse.report.html) | [94](https://circleci.com/gh/fariasf/drdt-test/2702/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Comment by akhan3051** [comment]
@fariasf, I have done the requested changes. Please take a look, when you have some time.

**Comment by fariasf** [comment]
#105 removes the whitespace issues

**Pull request closed without merging**